File: C++/level2/조이스틱.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

bool passFunction(vector<bool> pass){
    for(int i=0; i<pass.size(); i++)
        if(pass[i] == false)
            return false;
    return true;
}

int solution(string name) {
    int answer = 0;
    vector<bool> pass;

    for(int i=0; i<name.size(); i++){
        if(name[i] < 'L')
            answer +=((int)name[i]-65);
        else
            answer +=(91-(int)name[i]);
        if(name[i] == 'A')
             pass.push_back(true);
        else
            pass.push_back(false);
    }

    pass[0] = true;
    int left;
    int right;
    int index = 0;

    while(!passFunction(pass)){
        left = 0;
        right = 0;

        for(int i=0; i<pass.size(); i++)
            if(pass[(index+i)%pass.size()])
                right++;
            else break;

        for(int i=0; i<pass.size(); i++)
            if(pass[(pass.size()+index-i)%pass.size()])
                left++;
            else break;

        if(left >= right){
            answer += right;
            index = (index+right)%pass.size();
        }
        else{
            answer += left;
            index = (pass.size()+index-left)%pass.size();
        }
        pass[index] = true;
    }
    return answer;
}

File: JavaScript/level2/조이스틱.js
function passArray(pass){
    for(let i of pass)
        if(!i)
            return false;
    return true;
}

function solution(name) {
    var answer = 0;
    let SIZE = name.length;
    let pass = new Array(SIZE).fill(false);
    let left;
    let right;
    let current = 0;

    for(let i=0; i<SIZE; i++){
        answer += Math.min(91-name[i].charCodeAt(0), name[i].charCodeAt(0)-65)
        
        if(name[i] === 'A')
             pass[i] = true;
    }

    while(!passArray(pass)){
        left = 0;
        right = 0;
        
        for(let i=0; i<SIZE; i++)
            if(pass[(current+i)%SIZE])
                right++;
            else 
                break;

        for(let i=0; i<SIZE; i++)
            if(pass[(SIZE+current-i)%SIZE])
                left++;
            else 
                break;
        
        if(left >= right){
            answer += right;
            current = (current+right)%SIZE;
        }
        else{
            answer += left;
            current = (SIZE+current-left)%SIZE;
        }
        pass[current] = true;
    }
    
    return answer;
}